crypto/provider_core.c

```diff
@@ -1153,11 +1153,12 @@ static int provider_activate(OSSL_PROVIDER *prov, int lock, int upcalls)
         return -1;
     }
 
-    count = ++prov->activatecnt;
-    prov->flag_activated = 1;
+    if (CRYPTO_UP_REF(&prov->activatecnt, &count, prov->refcnt_lock) > 0) {
+        prov->flag_activated = 1;
```

Comment:
It seems like this won't fail right if CRYPTO_UP_REF fails?

Are we assuming normal store operations are atomic here (`flag_activated`)?

Reply:
> It seems like this won't fail right if CRYPTO_UP_REF fails?

Doesn't it? It looks ok to me? If CRYPTO_UP_REF fails then we assume `count` is unchanged at -1 and the body of the "if" is skipped. We go straight to the "unlocking" section below and return count (-1) to indicate an error.

> Are we assuming normal store operations are atomic here (flag_activated)?

I'm not too sure I understand your question. `flag_activated` is protected by the `flag_lock`. Reads/writes to it require that we hold the read/write lock. See the big comment about locking at the start of this file which explains how the locking works.